scipy/optimize/tests/test_quadratic_assignment.py

```diff
@@ -0,0 +1,203 @@
+        [0, 23, 0, 0, 0, 88, 0, 0, 0, 0, 0, 0],
+        [0, 0, 43, 0, 0, 0, 26, 0, 0, 0, 0, 0],
+        [0, 0, 0, 88, 0, 0, 0, 16, 0, 0, 0, 0],
+        [0, 0, 0, 0, 26, 0, 0, 0, 1, 0, 0, 0],
+        [0, 0, 0, 0, 0, 16, 0, 0, 0, 96, 0, 0],
+        [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 29, 0],
+        [0, 0, 0, 0, 0, 0, 0, 96, 0, 0, 0, 37],
+        [0, 0, 0, 0, 0, 0, 0, 0, 29, 0, 0, 0],
+        [0, 0, 0, 0, 0, 0, 0, 0, 0, 37, 0, 0],
+    ]
+    dist_matrix = [
+        [0, 36, 54, 26, 59, 72, 9, 34, 79, 17, 46, 95],
+        [36, 0, 73, 35, 90, 58, 30, 78, 35, 44, 79, 36],
+        [54, 73, 0, 21, 10, 97, 58, 66, 69, 61, 54, 63],
+        [26, 35, 21, 0, 93, 12, 46, 40, 37, 48, 68, 85],
+        [59, 90, 10, 93, 0, 64, 5, 29, 76, 16, 5, 76],
+        [72, 58, 97, 12, 64, 0, 96, 55, 38, 54, 0, 34],
+        [9, 30, 58, 46, 5, 96, 0, 83, 35, 11, 56, 37],
+        [34, 78, 66, 40, 29, 55, 83, 0, 44, 12, 15, 80],
+        [79, 35, 69, 37, 76, 38, 35, 44, 0, 64, 39, 33],
+        [17, 44, 61, 48, 16, 54, 11, 12, 64, 0, 70, 86],
+        [46, 79, 54, 68, 5, 0, 56, 15, 39, 70, 0, 18],
+        [95, 36, 63, 85, 76, 34, 37, 80, 33, 86, 18, 0],
+    ]
+
+    # check ofv with 100 random initializations
+    res = quadratic_assignment(
+        cost_matrix, dist_matrix, options={'init_weight': 0.5, 'init_n': 100}
+    )
+
+    assert 11156 <= res['score'] < 14500
+
+
+def test_quadratic_assignment_input_validation():
+    A = np.identity(2)
+    B = A
+
+    # Type Error Checks: making sure single value parameters are of
+    # correct type
+
+    with pytest.raises(TypeError):
+        quadratic_assignment(A, B, options={'init_n': 1.5})
+    with pytest.raises(ValueError):
+        quadratic_assignment(A, B, options={'init_n': -1})
+    with pytest.raises(ValueError):
+        quadratic_assignment(A, B, options={'init_weight': 2})
+    with pytest.raises(ValueError):
+        quadratic_assignment(A, B, options={'init': "random"})
+    with pytest.raises(TypeError):
+        quadratic_assignment(A, B, options={'maxiter': 1.5})
+    with pytest.raises(ValueError):
+        quadratic_assignment(A, B, options={'maxiter': -1})
+    with pytest.raises(TypeError):
+        quadratic_assignment(A, B, options={'shuffle_input': "hey"})
+    with pytest.raises(TypeError):
+        quadratic_assignment(A, B, options={'eps': -1})
+    with pytest.raises(TypeError):
+        quadratic_assignment(A, B, maximize="hey")
+    # test that cost and dist matrices of different sizes return error
+    with pytest.raises(ValueError):
+        quadratic_assignment(
+            np.random.random((3, 3)),
+            np.random.random((4, 4)),
+        )
+    # test that cost/dist matrices must be nonnegative
+    with pytest.raises(ValueError):
+        quadratic_assignment(
+            -1 * np.random.random((3, 3)),
+            np.random.random((3, 3))
+        )
+    # test that non square matrices return error
+    with pytest.raises(ValueError):
```

Comment:
```suggestion
    with pytest.raises(ValueError, match="'cost_matrix' must be square"):
```
Please add `match` strings for all errors. Two reasons: they help ensure that the error message matches what was intended., and they also help future reviewers understand what each test is for at a glance.

Re-ordering the tests in approximately the same order as you perform the checks in the code would be helpful, too.